### DISCUSSION TITLE
How to understand the parameters in the new line_width node?

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#274] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/36077492/158980386-55c52e6a-0489-4b97-9799-3ddcd797390e.png)

The scale and id_width are easy to understand. The rest are not intuitive. I did some tests. but still not sure how they work in different model

**Comment by eballeste** [comment]
in some of my tests I noticed that the normal threshold makes the lines disappear depending on the distance between the object and the camera

**Reply by eballeste** [reply]
but now im seeing the depth threshold setting and wondering if it wasn't the distance and it had more to do with the relationship between the angle of the normals and the camera view

**Reply by pragma37** [reply]
This is because the detection is based on contiguous pixels. So, in a beveled edge, the normal difference will be lower when the camera gets closer.

**Comment by pragma37** [comment, marked as answer]
The line detection works by detecting ID, depth, and normal discontinuities between nearby pixels.

The line width scale just multiplies the final result.

By default, the first channel (ID.x) is a unique number for each object, and the other 3 channels are set to 0.
So, by default, the ID detection will create contour lines at object boundaries.

The other channels can be used for custom boundaries:
![imagen](https://user-images.githubusercontent.com/36610999/159173565-068af1ae-bc0a-4396-8dd3-485952e8f08f.png)


The depth detection detects a line when the depth difference is larger than the depth threshold (expressed in world units).
The other parameters are used in a range mapping function.

So the line width will be equal to *depth width* at *depth threshold*, and it will be equal to *depth width* + *depth width range* at *depth threshold* + *depth threshold range*. 
![line-depth](https://user-images.githubusercontent.com/36610999/159173287-5fc99c22-2a9f-4952-8a35-9493b7b6634d.png)
So here:
- A: *depth* = 0.5 | *width* = 2.0
- B: *depth* = 0.5 + 2.0 | *width* = 2.0 + 6.0

The ranges can be negative too.

The normal detection detects a line when the normal difference is larger than the threshold (expressed in radians).
The ranges work the same as in the depth based detection.
